File: Ashutosh-Singh/BinarySearch/Capacity-To-Ship-Packages-Within-D-Days.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int findDays(vector<int> arr, int mid) {
        int weightPossible = 0;
        int days = 1;
        for (int i = 0; i < arr.size(); i++) {
            if (weightPossible + arr[i] > mid) {
                days++;
                weightPossible = arr[i];
            } else {
                weightPossible += arr[i];
            }
        }
        return days;
    }

    int shipWithinDays(vector<int>& weights, int days) {
        int start = *max_element(weights.begin(), weights.end());
        int end = 0;
        for (auto x : weights) {
            end += x;
        }

        while (start <= end) {
            int mid = start + (end - start) / 2;
            int NoOfDays = findDays(weights, mid);
            if (NoOfDays <= days) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return start;
    }
};

int main() {
    Solution solution;
    int n, days;

    cout << "Enter the number of packages: ";
    cin >> n;

    vector<int> weights(n);
    cout << "Enter the weights of the packages: ";
    for (int i = 0; i < n; i++) {
        cin >> weights[i];
    }

    cout << "Enter the number of days to deliver: ";
    cin >> days;

    int result = solution.shipWithinDays(weights, days);
    cout << "The minimum capacity of the ship to deliver within " << days << " days is: " << result << endl;

    return 0;
}

File: [Manya]/Leetcode/Binary-Search/capacity_to_ship_packages_within_d_days.java
import java.util.Scanner;

class Solution {
    public boolean isPossible(int[] weights, int capacity, int days) {
        int weight = 0, countDays = 1;

        for (int i = 0; i < weights.length; i++) {
            if (weight + weights[i] > capacity) {
                countDays++; 
                weight = weights[i]; 
                if (countDays > days) {
                    return false;
                }
            } else {
                weight += weights[i];
            }
        }
        return true;
    }

    public int shipWithinDays(int[] weights, int days) {
        int max = 0, min = 0;

        for (int weight : weights) {
            max += weight;  
            min = Math.max(min, weight); 
        }

        while (min <= max) {
            int mid = min + (max - min) / 2;

            if (isPossible(weights, mid, days)) {
                max = mid - 1;
            } else {
                min = mid + 1;
            }
        }
        return min;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter number of packages: ");
        int n = scanner.nextInt();
        int[] weights = new int[n];

        System.out.println("Enter package weights:");
        for (int i = 0; i < n; i++) {
            weights[i] = scanner.nextInt();
        }

        System.out.print("Enter number of days: ");
        int days = scanner.nextInt();

        Solution sol = new Solution();
        int minCapacity = sol.shipWithinDays(weights, days);
        System.out.println("Minimum ship capacity required: " + minCapacity);

        scanner.close();
    }
}
